### ISSUE TITLE
During Training model's loss doesn't converge

### ISSUE THREAD

**Issue by compressor1212** [facebookresearch/encodec#17]
## ❓ Questions

I have written my custom training loop with a reconstruction loss function. But the loss doesn't converge and gets bounced around 600 and 700 (in the case of my loss func) even after 3 to 4 epochs. Can you please explain why did that happen?

I first wanted to check the loss convergence of the model only for reconstruction loss without the discriminators.

loss function code (From SoundStream paper) : 
![image](https://user-images.githubusercontent.com/118351766/202157584-759f6827-49b2-4452-a547-e2f216b83d37.png)

```
def L_G_rec(x, G_x, eps=1e-4):
    L = 0
    sr = 16000
    for i in range(6, 12):
        s = 2 ** i
        alpha_s = (s / 2) ** 0.5
        melspec = MelSpectrogram(sample_rate=sr, n_fft=s, hop_length=s//4, center=True, pad_mode="reflect", power=2.0, norm="slaney", onesided=True, n_mels=8, mel_scale="htk")
        S_x = melspec(x)
        S_G_x = melspec(G_x)
        loss1 = (S_x - S_G_x).abs().mean()
        loss2 = alpha_s * (((torch.log(S_x.abs() + eps) - torch.log(S_G_x.abs() + eps)) ** 2).sum(dim=-2).add(eps).abs() ** 0.5).mean()
        loss = loss1 + loss2
        L = L + loss
    return L
```

Training loop code : 
```
model = EncodecModel._get_model(24)
. . . 
for epoch in range(0, EPOCH):
  print("EPOCH number = %d" % epoch)
  for i, input in enumerate(iter(dataloader)):
    with torch.autograd.set_detect_anomaly(True):

      recon_ouput = model(input)

      gen_loss = L_G_rec(input, recon_output)

      generator_optim.zero_grad()
      gen_loss.backward()
      generator_optim.step()
      if i % 50 == 0:
        print("Training generator loss = %d" % gen_loss)
      save_model(model, generator_optim)
```

What is the problem with the training loop? Do I need to include a discriminator for convergence?

**Comment by adefossez** [comment]
@compressor1212 while we are planning to release training code at a future date, we do not provide support for custom training code in the mean time.
